File: 20260727-game_multiplayer_websocket/src/match.js
// Match bookkeeping: scores, purses, which hull we are on, what each player last saw of
// the other. Pure state with no DOM and no renderer, so the eventual server can own it and
// main.js is left doing only presentation and flow.

import { createDesign, fitDesignToHull, cloneDesign } from './sim/ship.js';
import { HULLS } from './data/hulls.js';
import { ROUNDS, POINTS_TO_WIN, loserBonus } from './config.js';
import { makeRng, hashSeed } from './sim/rng.js';

export function createMatch(seed, fromRound = 0) {
  const match = {
    seed,
    roundIndex: fromRound,
    scores: [0, 0],
    designs: [createDesign(), createDesign()],
    scrap: [0, 0],
    lastSeen: [null, null], // what each player saw of the other after the last battle
    lastLoser: null,
    windTo: 0,
  };
  // Starting at a later round has to hand over the scrap the skipped rounds would have
  // granted, or the big hulls turn up nearly empty.
  for (let r = 0; r < fromRound; r++) {
    match.scrap[0] += ROUNDS[r].scrap;
    match.scrap[1] += ROUNDS[r].scrap;
  }
  return match;
}

export const hullIndexOf = (match) => ROUNDS[match.roundIndex].hull;
export const hullOf = (match) => HULLS[hullIndexOf(match)];
export const roundOf = (match) => ROUNDS[match.roundIndex];

// Roll the wind, move both ships into this round's hull, and pay out.
export function beginRound(match) {
  const hullIndex = hullIndexOf(match);
  match.windTo = makeRng(hashSeed(match.seed, match.roundIndex, 77)).next() * Math.PI * 2;
  const bonus = match.lastLoser === null ? 0 : loserBonus(match.roundIndex);
  for (let i = 0; i < 2; i++) {
    fitDesignToHull(match.designs[i], hullIndex);
    match.scrap[i] += roundOf(match).scrap + (match.lastLoser === i ? bonus : 0);
  }
  return { hullIndex, bonus };
}

export function battleSeed(match) {
  return hashSeed(match.seed, match.roundIndex, 4242);
}

// What became of a ship, for the result screen. A running log says what happened; this says which
// decision was wrong, which is the thing a player watching a battle they cannot steer actually needs.
// Reads only the battle's own state, so it stays pure.
export function roundSummary(ship) {
  let mastsHad = 0;
  let mastsLeft = 0;
  let handsHad = 0;
  for (const cell of ship.cells) {
    handsHad += cell.crewSupply;
    if (cell.id === 'mast') {
      mastsHad++;
      if (cell.alive) mastsLeft++;
    }
  }
  let gunsLeft = 0;
  let firing = 0;
  for (const gun of ship.guns) {
    if (gun.cell.alive) gunsLeft++;
    if (gun.manned) firing++;
  }
  return {
    cellsHad: ship.cells.length,
    cellsLeft: ship.aliveCells,
    gunsHad: ship.guns.length,
    gunsLeft,
    firing,
    mastsHad,
    mastsLeft,
    handsHad,
    hands: ship.crew,
    powder: ship.magazines,
  };
}

export function offerSeed(match, player) {
  return hashSeed(match.seed, match.roundIndex, player, 913);
}

// Award the point and let both players see what they were up against.
export function recordResult(match, winner) {
  if (winner === null) {
    match.lastLoser = null;
  } else {
    match.scores[winner]++;
    match.lastLoser = 1 - winner;
  }
  const hullIndex = hullIndexOf(match);
  for (let i = 0; i < 2; i++) {
    match.lastSeen[i] = { design: cloneDesign(match.designs[i]), hullIndex };
  }
}

export function intelFor(match, player) {
  const seen = match.lastSeen[1 - player];
  return seen ? { design: seen.design, hullIndex: seen.hullIndex } : null;
}

export function isMatchOver(match) {
  return (
    match.scores[0] >= POINTS_TO_WIN ||
    match.scores[1] >= POINTS_TO_WIN ||
    match.roundIndex >= ROUNDS.length - 1
  );
}

File: 20260815-game_multiplayer_port_raylib/native/game/match.cpp
#include "match.h"
#include "sim/geometry.h"
#include "sim/rng.h"
#include <algorithm>
#include <cmath>
namespace broadside::game {
const std::vector<RoundInfo> &rounds() {
    static const std::vector<RoundInfo> r = {{0, 34, 40}, {1, 32, 26}, {2, 42, 26}, {3, 46, 28}, {4, 56, 30}};
    return r;
}
Match makeMatch(std::uint32_t seed, int players) {
    Match m;
    m.seed = seed;
    m.players = players;
    m.scores.assign(players, 0);
    m.scrap.assign(players, 0);
    m.designs.resize(players, sim::makeDesign());
    m.damage.assign(players, 0);
    return m;
}
void beginRound(Match &m) {
    auto r = rounds()[m.round];
    sim::Rng rng(sim::hashSeed({m.seed, static_cast<std::uint32_t>(m.round), 0x9e3779b9u}));
    m.wind = rng.range(0, sim::Pi * 2);
    for (auto &s : m.scrap)
        s += r.scrap;
    for (auto &d : m.designs)
        sim::fitDesignToHull(d, r.hull);
}
void recordResult(Match &m, const sim::Result &r) {
    if (r.winner >= 0 && r.winner < m.players)
        m.scores[r.winner]++;
    for (int place = 1; place < static_cast<int>(r.placing.size()); ++place)
        m.scrap[r.placing[place]] += static_cast<int>(
            std::lround(rounds()[m.round].scrap * .45f * static_cast<float>(place) / (m.players - 1)));
    if (winner(m) >= 0) {
        m.over = true;
        return;
    }
    if (++m.round >= static_cast<int>(rounds().size()))
        m.over = true;
}
int winner(const Match &m) {
    for (int i = 0; i < m.players; ++i)
        if (m.scores[i] >= 3)
            return i;
    if (!m.over && m.round < static_cast<int>(rounds().size()))
        return -1;
    int leader = -1, best = -1;
    bool tied = false;
    for (int i = 0; i < m.players; ++i) {
        if (m.scores[i] > best) {
            best = m.scores[i];
            leader = i;
            tied = false;
        } else if (m.scores[i] == best)
            tied = true;
    }
    return tied ? -1 : leader;
}
} // namespace broadside::game
